### PR TITLE
Feature/improvements

### PR DESCRIPTION
**Pull request by rburtnykx**
FIXED: issue when std output is redirected (sys.stdout.encoding is None)
ADDED: use compile base directory from gcov json; Here: https://gcc.gnu.org/onlinedocs/gcc/Invoking-Gcov.html#Invoking-Gcov 
{
  "**current_working_directory": current_working_directory**,
  "data_file": data_file,
...

ADDED: print statistic feature; 
ADDED: source existence check

### CODE REVIEW

**Comment by RPGillespie6** [comment]
Could you elaborate on what the `sys.stdout.encoding` issue is and how to reproduce?

I need to check if `current_working_directory` has been there since GCC 9.0, if so we can make this change. We should still allow the user to override it though, to retain backward compatibility.

**Comment by rburtnykx** [comment]
Yes, sure. In case of redirected output, f.e. to file, `sys.stdout.encoding` is None. So error will be smth like next:

```
Traceback (most recent call last):
...
    intermediate_json = json.loads(line.decode(sys.stdout.encoding))
TypeError: decode() argument 1 must be string, not None
```
So to reproduce step is next:
`python fastcov.py > test.txt`
Script should run getGcovCoverage.

About GCC 9.0. Sure. Unfortunately, I was not able to check it for gcc 9.0, only 9.1+, but it was mentioned in all docs for gcc9.
About backward compatibility. How will be better to do that? I see 2 solutions:
1. adding additional cmdline parameter which will ingore -c, (like --gcov-determine-compile-dir) 
2. If -c is not set, use current_work_dir json field from gcov

**Comment by RPGillespie6** [comment]
Ah, I see. Good catch, I had not thought to try that. Thanks for adding the redirect (and other) test cases to `test/functional/run_all.sh`

> About backward compatibility. How will be better to do that? I see 2 solutions:

I think the second solution is better. If `-c` is not set, use gcov field. If `-c` is set, override gcov field with what was set.

I will take a closer look at this PR and see if I have any other suggestions, but so far I like these improvements.

**Comment by rburtnykx** [comment]
Great, thanks. So i will definitely leave here this 2 items: redirect fix and base path using gcov output. Even if gcov has no that attribute in json, f.e. for 9.0.x version, it will behave the same after changes.

Other improvements are:
1. Exclude using glob. 
2. Dump short statistic at the end. That is smth like in lcov, sometimes useful to see that. If this is OK, i think i need to add also other info, like files, branches, etc.
3. Source existence check. 

Please, let me know if smth better to not include so i will remove that.

File: fastcov.py [hunk 1]
```diff
@@ -368,6 +374,12 @@ def scanExclusionMarkers(fastcov_json, jobs, exclude_branches_sw, include_branch
     for t in threads:
         t.join()
 
+def validateSources(fastcov_json):
+    logging.info("Checking if all sources exist")
+    for source in fastcov_json["sources"].keys():
+        if not os.path.exists(source):
+            logging.error("Cannot found '{}'".format(source))
```
Comment (RPGillespie6) [inline fastcov.py:381]: Small typo, should be `Cannot find`

Comment (rburtnykx) [reply fastcov.py]: Fixed, thanks

**Comment by rburtnykx** [comment]
Found and fixed a little issue with default file name. When --lcov parameter is set, default file name is coverage.json. Changed to  coverage.info if --lcov is set, coverage.json otherwise

**Comment by RPGillespie6** [comment]
Is this ready for merge now?

**Comment by rburtnykx** [comment]
I think yes

**Comment by RPGillespie6** [comment]
Alright, I'm going to merge it. Thanks for contributing!

**Comment by RPGillespie6** [comment]
I'll probably make a follow on PR that bumps the fastcov version, adds a few more tests, and other preparations for `v1.8`

**Comment by rburtnykx** [comment]
Awesome, thanks for taking my changes into branch.

**Pull request merged**